// @wind/wind-ui/lib/timeline/TimelineItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

var TimelineItem = /*#__PURE__*/function (_React$Component) {
  _inherits(TimelineItem, _React$Component);

  var _super = _createSuper(TimelineItem);

  function TimelineItem() {
    _classCallCheck(this, TimelineItem);

    return _super.apply(this, arguments);
  }

  _createClass(TimelineItem, [{
    key: "render",
    value: function render() {
      var _classNames, _classNames2, _classNames3;

      var _a = this.props,
          prefixCls = _a.prefixCls,
          className = _a.className,
          _a$color = _a.color,
          color = _a$color === void 0 ? '' : _a$color,
          last = _a.last,
          children = _a.children,
          pending = _a.pending,
          dot = _a.dot,
          title = _a.title,
          titlePos = _a.titlePos,
          restProps = __rest(_a, ["prefixCls", "className", "color", "last", "children", "pending", "dot", "title", "titlePos"]); //const tPos =  titlePos||'left';


      var itemClassName = (0, _classnames["default"])((_classNames = {}, _defineProperty(_classNames, "".concat(prefixCls, "-item"), true), _defineProperty(_classNames, "".concat(prefixCls, "-item-last"), last), _defineProperty(_classNames, "".concat(prefixCls, "-item-pending"), pending), _defineProperty(_classNames, "".concat(prefixCls, "-item-title-").concat(titlePos || 'left'), true), _classNames), className);
      var dotClassName = (0, _classnames["default"])((_classNames2 = {}, _defineProperty(_classNames2, "".concat(prefixCls, "-item-head"), true), _defineProperty(_classNames2, "".concat(prefixCls, "-item-head-custom"), dot), _defineProperty(_classNames2, "".concat(prefixCls, "-item-head-").concat(color), !!color), _classNames2));
      var titleClassName = (0, _classnames["default"])((_classNames3 = {}, _defineProperty(_classNames3, "".concat(prefixCls, "-item-title"), true), _defineProperty(_classNames3, "".concat(prefixCls, "-item-empty"), !title), _classNames3));
      return /*#__PURE__*/_react["default"].createElement("li", _extends({}, restProps, {
        className: itemClassName
      }), titlePos === 'right' ? null : /*#__PURE__*/_react["default"].createElement("div", {
        className: titleClassName
      }, title), /*#__PURE__*/_react["default"].createElement("div", {
        className: "".concat(prefixCls, "-item-tail")
      }), /*#__PURE__*/_react["default"].createElement("div", {
        className: "".concat(prefixCls, "-item-line")
      }, /*#__PURE__*/_react["default"].createElement("div", {
        className: dotClassName,
        style: {
          borderColor: /blue|red|green/.test(color) ? undefined : color
        }
      }, dot)), titlePos === 'right' ? /*#__PURE__*/_react["default"].createElement("div", {
        className: titleClassName
      }, title) : null, /*#__PURE__*/_react["default"].createElement("div", {
        className: "".concat(prefixCls, "-item-content")
      }, children));
    }
  }]);

  return TimelineItem;
}(_react["default"].Component);

exports["default"] = TimelineItem;
TimelineItem.defaultProps = {
  prefixCls: 'w-timeline',
  color: '',
  last: false,
  pending: false
};
