// @wind/wind-ui/lib/notification/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _rcNotification = _interopRequireDefault(require("rc-notification"));

var _classnames = _interopRequireDefault(require("classnames"));

var _CheckCircleO = _interopRequireDefault(require("@wind/icons/lib/icons/CheckCircleO"));

var _InfoCircleO = _interopRequireDefault(require("@wind/icons/lib/icons/InfoCircleO"));

var _ExclamationO = _interopRequireDefault(require("@wind/icons/lib/icons/ExclamationO"));

var _CloseCircleO = _interopRequireDefault(require("@wind/icons/lib/icons/CloseCircleO"));

var _CloseO = _interopRequireDefault(require("@wind/icons/lib/icons/CloseO"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var notificationInstance = {};
var defaultDuration = 4.5;
var defaultTop = 24;
var defaultBottom = 24;
var defaultPlacement = 'topRight';
var defaultGetContainer;

function setNotificationConfig(options) {
  var duration = options.duration,
      placement = options.placement,
      bottom = options.bottom,
      top = options.top,
      getContainer = options.getContainer;
  var needRemove = false;

  if (duration !== undefined) {
    defaultDuration = duration;
  }

  if (placement !== undefined) {
    defaultPlacement = placement;
  }

  if (bottom !== undefined) {
    defaultBottom = bottom;
    needRemove = true;
  }

  if (top !== undefined) {
    defaultTop = top;
    needRemove = true;
  }

  if (getContainer !== undefined) {
    defaultGetContainer = getContainer;
    needRemove = true;
  }

  if (needRemove) {
    removeInstance();
  }
} // 部分配置变化时需要重新生成 Instance，故不建议经常调用 config 方法。


function removeInstance() {
  // 重新生成 Instance
  Object.keys(notificationInstance).forEach(function (cacheKey) {
    return notificationInstance[cacheKey] = null;
  });
}

function getPlacementStyle(placement) {
  var top = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : defaultTop;
  var bottom = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : defaultBottom;
  var style;

  switch (placement) {
    case 'topLeft':
      style = {
        left: 0,
        top: top,
        bottom: 'auto'
      };
      break;

    case 'topRight':
      style = {
        right: 0,
        top: top,
        bottom: 'auto'
      };
      break;

    case 'bottomLeft':
      style = {
        left: 0,
        top: 'auto',
        bottom: bottom
      };
      break;

    default:
      style = {
        right: 0,
        top: 'auto',
        bottom: bottom
      };
      break;
  }

  return style;
}

function getNotificationInstance(_ref, callback) {
  var prefixCls = _ref.prefixCls,
      _ref$placement = _ref.placement,
      placement = _ref$placement === void 0 ? defaultPlacement : _ref$placement,
      _ref$getContainer = _ref.getContainer,
      getContainer = _ref$getContainer === void 0 ? defaultGetContainer : _ref$getContainer,
      top = _ref.top,
      bottom = _ref.bottom;
  var cacheKey = "".concat(prefixCls, "-").concat(placement);

  if (notificationInstance[cacheKey]) {
    callback(notificationInstance[cacheKey]);
    return;
  }

  _rcNotification["default"].newInstance({
    prefixCls: prefixCls,
    className: "".concat(prefixCls, "-").concat(placement),
    style: getPlacementStyle(placement, top, bottom),
    getContainer: getContainer,
    closeIcon: /*#__PURE__*/_react["default"].createElement(_CloseO["default"], null)
  }, function (notification) {
    notificationInstance[cacheKey] = notification;
    callback(notification);
  });
}

var typeToIcon = {
  success: /*#__PURE__*/_react["default"].createElement(_CheckCircleO["default"], {
    className: "w-notification-notice-icon w-notification-notice-icon-success"
  }),
  info: /*#__PURE__*/_react["default"].createElement(_InfoCircleO["default"], {
    className: "w-notification-notice-icon w-notification-notice-icon-info"
  }),
  error: /*#__PURE__*/_react["default"].createElement(_CloseCircleO["default"], {
    className: "w-notification-notice-icon w-notification-notice-icon-error"
  }),
  warning: /*#__PURE__*/_react["default"].createElement(_ExclamationO["default"], {
    className: "w-notification-notice-icon w-notification-notice-icon-warning"
  })
};

function notice(args) {
  var outerPrefixCls = args.prefixCls || 'w-notification';
  var prefixCls = "".concat(outerPrefixCls, "-notice");
  var duration = args.duration === undefined ? defaultDuration : args.duration;
  var iconNode = null;

  if (args.icon) {
    iconNode = /*#__PURE__*/_react["default"].createElement("span", {
      className: "".concat(prefixCls, "-icon")
    }, args.icon);
  } else if (args.type) {
    var iconType = typeToIcon[args.type];
    iconNode = iconType;
  }

  var autoMarginTag = !args.description && iconNode ? /*#__PURE__*/_react["default"].createElement("span", {
    className: "".concat(prefixCls, "-message-single-line-auto-margin")
  }) : null;
  var placement = args.placement;
  getNotificationInstance({
    prefixCls: outerPrefixCls,
    placement: placement
  }, function (notification) {
    notification.notice({
      content: /*#__PURE__*/_react["default"].createElement("div", {
        className: (0, _classnames["default"])(iconNode ? "".concat(prefixCls, "-with-icon") : '', !args.description ? "".concat(prefixCls, "-title-only") : '')
      }, iconNode, /*#__PURE__*/_react["default"].createElement("div", {
        className: "".concat(prefixCls, "-message")
      }, autoMarginTag, args.message), /*#__PURE__*/_react["default"].createElement("div", {
        className: "".concat(prefixCls, "-description")
      }, args.description), args.btn ? /*#__PURE__*/_react["default"].createElement("span", {
        className: "".concat(prefixCls, "-btn")
      }, args.btn) : null),
      duration: duration,
      closable: true,
      onClose: args.onClose,
      onClick: args.onClick,
      key: args.key,
      style: args.style || {},
      className: args.className
    });
  });
}

var api = {
  open: notice,
  close: function close(key) {
    Object.keys(notificationInstance).forEach(function (cacheKey) {
      return notificationInstance[cacheKey].removeNotice(key);
    });
  },
  config: setNotificationConfig,
  destroy: function destroy() {
    Object.keys(notificationInstance).forEach(function (cacheKey) {
      notificationInstance[cacheKey].destroy();
      delete notificationInstance[cacheKey];
    });
  }
};
['success', 'info', 'warning', 'error'].forEach(function (type) {
  api[type] = function (args) {
    return api.open(_extends(_extends({}, args), {
      type: type
    }));
  };
});
api.warn = api.warning;
var _default = api;
exports["default"] = _default;
