// @wind/wind-ui/es/modal/confirm.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React from 'react';
import ReactDOM from 'react-dom';
import classNames from 'classnames';
import QuestionCircleF from "@wind/icons/es/icons/QuestionCircleF";
import Modal from './Modal';
import ActionButton from './ActionButton';
import { getConfirmLocale } from './locale';
export default function confirm(config) {
  var props = _extends({
    icon: /*#__PURE__*/React.createElement(QuestionCircleF, null)
  }, config);

  var prefixCls = props.prefixCls || 'w-confirm';
  var div = document.createElement('div');
  document.body.appendChild(div);
  var width = props.width || 416;
  var style = props.style || {};
  var okType = props.okType || 'primary'; // 默认为 false，保持旧版默认行为

  var maskClosable = props.maskClosable === undefined ? false : props.maskClosable; // 默认为 true，保持向下兼容

  if (!('okCancel' in props)) {
    props.okCancel = true;
  }

  var runtimeLocale = getConfirmLocale();
  props.okText = props.okText || (props.okCancel ? runtimeLocale.okText : runtimeLocale.justOkText);
  props.cancelText = props.cancelText || runtimeLocale.cancelText;

  function close() {
    var unmountResult = ReactDOM.unmountComponentAtNode(div);

    if (unmountResult && div.parentNode) {
      div.parentNode.removeChild(div);
    }

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var triggerCancel = args && args.length && args.some(function (param) {
      return param && param.triggerCancel;
    });

    if (props.onCancel && triggerCancel) {
      props.onCancel.apply(props, args);
    }
  }

  var body = /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-body")
  }, props.icon, /*#__PURE__*/React.createElement("span", {
    className: "".concat(prefixCls, "-title")
  }, props.title), /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-content")
  }, props.content));
  var footer = null;

  if (props.okCancel) {
    footer = /*#__PURE__*/React.createElement("div", {
      className: "".concat(prefixCls, "-btns")
    }, /*#__PURE__*/React.createElement(ActionButton, {
      actionFn: props.onCancel,
      closeModal: close
    }, props.cancelText), /*#__PURE__*/React.createElement(ActionButton, {
      type: okType,
      actionFn: props.onOk,
      closeModal: close,
      autoFocus: true
    }, props.okText));
  } else {
    footer = /*#__PURE__*/React.createElement("div", {
      className: "".concat(prefixCls, "-btns")
    }, /*#__PURE__*/React.createElement(ActionButton, {
      type: okType,
      actionFn: props.onOk,
      closeModal: close,
      autoFocus: true
    }, props.okText));
  }

  var classString = classNames(prefixCls, _defineProperty({}, "".concat(prefixCls, "-").concat(props.type), true), props.className);
  ReactDOM.render( /*#__PURE__*/React.createElement(Modal, {
    className: classString,
    onCancel: close.bind(this, {
      triggerCancel: true
    }),
    visible: true,
    title: "",
    transitionName: "zoom",
    footer: "",
    maskTransitionName: "fade",
    maskClosable: maskClosable,
    style: style,
    width: width
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(prefixCls, "-body-wrapper")
  }, body, " ", footer)), div);
  return {
    destroy: close
  };
}
