// @wind/wind-ui/lib/table/HorizontalTable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _FrownO = _interopRequireDefault(require("@wind/icons/lib/icons/FrownO"));

var _spin = _interopRequireDefault(require("../spin"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var defaultLocale = {
  emptyText: /*#__PURE__*/_react["default"].createElement("span", null, /*#__PURE__*/_react["default"].createElement(_FrownO["default"], null), "\u6682\u65E0\u6570\u636E")
};

function getRowKey(key, dataIndex, isTitle) {
  var ret = key == null ? dataIndex : key;

  if (ret != null) {
    return isTitle ? "".concat(ret, "-title") : ret;
  }

  return undefined;
}

var HorizontalTable = /*#__PURE__*/function (_React$Component) {
  _inherits(HorizontalTable, _React$Component);

  var _super = _createSuper(HorizontalTable);

  function HorizontalTable(props) {
    var _this;

    _classCallCheck(this, HorizontalTable);

    _this = _super.call(this, props);

    _this.generateTable = function () {
      var _this$props = _this.props,
          rows = _this$props.rows,
          dataSource = _this$props.dataSource,
          prefixCls = _this$props.prefixCls;

      var locale = _this.getLocale();

      if (dataSource) {
        var trs = [];
        rows && rows.forEach(function (row, index) {
          var innerRow = [];
          var len = row.length;
          row.forEach(function (item, index) {
            var _classNames;

            var dataIndex = item.dataIndex,
                key = item.key,
                titleWidth = item.titleWidth,
                contentWidth = item.contentWidth,
                render = item.render,
                title = item.title,
                className = item.className,
                colSpan = item.colSpan,
                titleAlign = item.titleAlign,
                contentAlign = item.contentAlign;
            var titleStyle = {
              width: titleWidth,
              textAlign: titleAlign
            };
            var contentStyle = {
              width: contentWidth,
              textAlign: contentAlign
            };
            innerRow.push( /*#__PURE__*/_react["default"].createElement("td", {
              key: getRowKey(key, dataIndex, true),
              className: (0, _classnames["default"])(className, "".concat(prefixCls, "-item-title")),
              style: titleStyle
            }, title));
            innerRow.push( /*#__PURE__*/_react["default"].createElement("td", {
              key: getRowKey(key, dataIndex, false),
              colSpan: colSpan,
              className: (0, _classnames["default"])(className, (_classNames = {}, _defineProperty(_classNames, "".concat(prefixCls, "-last-column"), index === len - 1), _defineProperty(_classNames, "".concat(prefixCls, "-item-content"), true), _classNames)),
              style: contentStyle
            }, typeof render === 'function' ? render(dataSource[dataIndex], dataSource, index) : dataSource[dataIndex]));
          });
          trs.push( /*#__PURE__*/_react["default"].createElement("tr", {
            key: index,
            className: "".concat(prefixCls, "-row ").concat(prefixCls, "-row-").concat(index % 2 === 0 ? 'odd' : 'even')
          }, innerRow));
        });
        return /*#__PURE__*/_react["default"].createElement("table", null, /*#__PURE__*/_react["default"].createElement("tbody", {
          className: "".concat(prefixCls, "-tbody")
        }, trs));
      }

      return locale.emptyText;
    };

    return _this;
  }

  _createClass(HorizontalTable, [{
    key: "getLocale",
    value: function getLocale() {
      var locale = {};

      if (this.context.antLocale && this.context.antLocale.Table) {
        locale = this.context.antLocale.Table;
      }

      return _extends(_extends(_extends({}, defaultLocale), locale), this.props.locale);
    }
  }, {
    key: "render",
    value: function render() {
      var _classNames2;

      var _this$props2 = this.props,
          prefixCls = _this$props2.prefixCls,
          className = _this$props2.className,
          style = _this$props2.style,
          title = _this$props2.title,
          striped = _this$props2.striped,
          bordered = _this$props2.bordered,
          size = _this$props2.size;
      var loading = this.props.loading;

      if (typeof loading === 'boolean') {
        loading = {
          spinning: loading
        };
      }

      return /*#__PURE__*/_react["default"].createElement("div", {
        className: (0, _classnames["default"])("".concat(prefixCls, "-horizontal ").concat(prefixCls), className, (_classNames2 = {}, _defineProperty(_classNames2, "".concat(prefixCls, "-without-column-header"), true), _defineProperty(_classNames2, "".concat(prefixCls, "-striped"), striped), _defineProperty(_classNames2, "".concat(prefixCls, "-bordered-").concat(bordered), !!bordered), _defineProperty(_classNames2, "".concat(prefixCls, "-").concat(size), !!size), _classNames2)),
        style: style
      }, title ? /*#__PURE__*/_react["default"].createElement("div", {
        className: "".concat(prefixCls, "-title")
      }, title) : null, /*#__PURE__*/_react["default"].createElement(_spin["default"], loading, /*#__PURE__*/_react["default"].createElement("div", {
        className: "".concat(prefixCls, "-body")
      }, this.generateTable())));
    }
  }]);

  return HorizontalTable;
}(_react["default"].Component);

exports["default"] = HorizontalTable;
HorizontalTable.defaultProps = {
  dataSource: {},
  prefixCls: 'w-table',
  size: 'default',
  bordered: 'solid',
  loading: false,
  locale: {},
  striped: true
};
HorizontalTable.contextTypes = {
  antLocale: _propTypes["default"].object
};
