// @wind/wind-ui/lib/modal/Modal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _rcDialog = _interopRequireDefault(require("rc-dialog"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _addEventListener = _interopRequireDefault(require("rc-util/lib/Dom/addEventListener"));

var _CloseO = _interopRequireDefault(require("@wind/icons/lib/icons/CloseO"));

var _button = _interopRequireDefault(require("../button"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var mousePosition;
var mousePositionEventBinded;

var Modal = /*#__PURE__*/function (_React$Component) {
  _inherits(Modal, _React$Component);

  var _super = _createSuper(Modal);

  function Modal() {
    var _this;

    _classCallCheck(this, Modal);

    _this = _super.apply(this, arguments);

    _this.handleCancel = function (e) {
      var onCancel = _this.props.onCancel;

      if (onCancel) {
        onCancel(e);
      }
    };

    _this.handleOk = function (e) {
      var onOk = _this.props.onOk;

      if (onOk) {
        onOk(e);
      }
    };

    return _this;
  }

  _createClass(Modal, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (mousePositionEventBinded) {
        return;
      } // 只有点击事件支持从鼠标位置动画展开


      (0, _addEventListener["default"])(document.documentElement, 'click', function (e) {
        mousePosition = {
          x: e.pageX,
          y: e.pageY
        }; // 100ms 内发生过点击事件，则从点击位置动画展示
        // 否则直接 zoom 展示
        // 这样可以兼容非点击方式展开

        setTimeout(function () {
          return mousePosition = null;
        }, 100);
      });
      mousePositionEventBinded = true;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          okText = _this$props.okText,
          okType = _this$props.okType,
          cancelText = _this$props.cancelText,
          confirmLoading = _this$props.confirmLoading,
          footer = _this$props.footer,
          visible = _this$props.visible,
          className = _this$props.className,
          type = _this$props.type,
          footerClassName = _this$props.footerClassName,
          prefixCls = _this$props.prefixCls;
      var classes = (0, _classnames["default"])(className, _defineProperty({}, "".concat(Modal.defaultProps.prefixCls, "-").concat(type), type));

      if (this.context.antLocale && this.context.antLocale.Modal) {
        okText = okText || this.context.antLocale.Modal.okText;
        cancelText = cancelText || this.context.antLocale.Modal.cancelText;
      }

      var defaultFooter = /*#__PURE__*/_react["default"].createElement("div", {
        className: footerClassName
      }, /*#__PURE__*/_react["default"].createElement(_button["default"], {
        key: "cancel" //size="large"
        ,
        onClick: this.handleCancel
      }, cancelText || '取消'), /*#__PURE__*/_react["default"].createElement(_button["default"], {
        key: "confirm",
        type: okType || "primary" //primary
        //size="large"
        ,
        loading: confirmLoading,
        onClick: this.handleOk
      }, okText || '确定'));

      return /*#__PURE__*/_react["default"].createElement(_rcDialog["default"], _extends({
        onClose: this.handleCancel,
        closeIcon: /*#__PURE__*/_react["default"].createElement(_CloseO["default"], {
          className: "".concat(prefixCls, "-close-x")
        }),
        footer: footer === undefined ? defaultFooter : footer
      }, this.props, {
        className: classes,
        visible: visible,
        mousePosition: mousePosition
      }));
    }
  }]);

  return Modal;
}(_react["default"].Component);

exports["default"] = Modal;
Modal.defaultProps = {
  prefixCls: 'w-modal',
  width: 520,
  transitionName: 'zoom',
  maskTransitionName: 'fade',
  confirmLoading: false,
  visible: false
};
Modal.propTypes = {
  prefixCls: _propTypes["default"].string,
  onOk: _propTypes["default"].func,
  onCancel: _propTypes["default"].func,
  okText: _propTypes["default"].node,
  okType: _propTypes["default"].node,
  cancelText: _propTypes["default"].node,
  width: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),
  confirmLoading: _propTypes["default"].bool,
  visible: _propTypes["default"].bool,
  align: _propTypes["default"].object,
  footer: _propTypes["default"].node,
  title: _propTypes["default"].node,
  closable: _propTypes["default"].bool,
  footerClassName: _propTypes["default"].string
};
Modal.contextTypes = {
  antLocale: _propTypes["default"].object
};
