// @wind/wind-ui/lib/locale-provider/nl_NL.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _moment = _interopRequireDefault(require("moment"));

var _nl_NL = _interopRequireDefault(require("rc-pagination/lib/locale/nl_NL"));

var _nl_NL2 = _interopRequireDefault(require("../date-picker/locale/nl_NL"));

var _nl_NL3 = _interopRequireDefault(require("../time-picker/locale/nl_NL"));

var _nl_NL4 = _interopRequireDefault(require("../calendar/locale/nl_NL"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

_moment["default"].locale('nl');

var _default = {
  locale: 'nl',
  Pagination: _nl_NL["default"],
  DatePicker: _nl_NL2["default"],
  TimePicker: _nl_NL3["default"],
  Calendar: _nl_NL4["default"],
  Table: {
    filterTitle: 'Filter Menu',
    filterConfirm: 'OK',
    filterReset: 'Reset',
    emptyText: 'Geen gegevens'
  },
  Modal: {
    okText: 'OK',
    cancelText: 'Annuleren',
    justOkText: 'OK'
  },
  Popconfirm: {
    okText: 'OK',
    cancelText: 'Annuleren'
  },
  Transfer: {
    notFoundContent: 'Niet gevonden',
    searchPlaceholder: 'Zoeken',
    itemUnit: 'item',
    itemsUnit: 'items'
  },
  Select: {
    notFoundContent: 'Niet gevonden'
  }
};
exports["default"] = _default;
