// @wind/wind-ui/lib/locale-provider/tr_TR.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _moment = _interopRequireDefault(require("moment"));

var _en_US = _interopRequireDefault(require("rc-pagination/lib/locale/en_US"));

var _tr_TR = _interopRequireDefault(require("../date-picker/locale/tr_TR"));

var _tr_TR2 = _interopRequireDefault(require("../time-picker/locale/tr_TR"));

var _en_US2 = _interopRequireDefault(require("../calendar/locale/en_US"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

_moment["default"].locale('tr');

var _default = {
  locale: 'tr',
  Pagination: _en_US["default"],
  DatePicker: _tr_TR["default"],
  TimePicker: _tr_TR2["default"],
  Calendar: _en_US2["default"],
  Table: {
    filterTitle: 'Menü Filtrele',
    filterConfirm: 'Tamam',
    filterReset: 'Sıfırla',
    emptyText: 'Veri Yok',
    selectAll: 'Hepsini Seç',
    selectInvert: 'Tersini Seç'
  },
  Modal: {
    okText: 'Tamam',
    cancelText: 'İptal',
    justOkText: 'Tamam'
  },
  Popconfirm: {
    okText: 'Tamam',
    cancelText: 'İptal'
  },
  Transfer: {
    notFoundContent: 'Bulunamadı',
    searchPlaceholder: 'Arama',
    itemUnit: 'Öğe',
    itemsUnit: 'Öğeler'
  },
  Select: {
    notFoundContent: 'Bulunamadı'
  },
  Upload: {
    uploading: 'Yükleniyor...',
    removeFile: "Dosyay\u0131 kald\u0131r",
    uploadError: 'Yükleme Hatası',
    previewFile: "Dosyay\u0131 \xD6nizle"
  }
};
exports["default"] = _default;
